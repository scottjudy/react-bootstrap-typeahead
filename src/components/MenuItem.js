// @flow

import cx from 'classnames';
import React, { type Node } from 'react';

import { withItem } from '../behaviors/item';

import type { EventHandler } from '../types';

type MenuItemProps = {
  active?: boolean,
  children?: Node,
  className?: string,
  disabled?: boolean,
  href?: string,
  onClick?: EventHandler<HTMLElement>,
};

const BaseMenuItem = React.forwardRef<MenuItemProps, ?HTMLElement>((
  { active, children, className, disabled, onClick, ...props },
  ref
) => {
  return (
    <a
      {...props}
      className={cx('dropdown-item', { active, disabled }, className)}
      href={props.href || '#'}
      onClick={(e: SyntheticEvent<HTMLElement>) => {
        e.preventDefault();
        !disabled && onClick && onClick(e);
      }}
      ref={ref}>
      {children}
    </a>
  );
});

export { BaseMenuItem };
export default withItem(BaseMenuItem);
