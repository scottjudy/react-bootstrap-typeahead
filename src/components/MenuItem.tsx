import cx from 'classnames';
import React, { forwardRef, HTMLProps, MouseEvent } from 'react';

import { useItem, UseItemProps } from '../behaviors/item';

export interface MenuItemProps extends HTMLProps<HTMLAnchorElement> {
  active?: boolean;
  disabled?: boolean;
}

export const BaseMenuItem = forwardRef<HTMLAnchorElement, MenuItemProps>(
  ({ active, children, className, disabled, onClick, ...props }, ref) => {
    return (
      <a
        {...props}
        className={cx('dropdown-item', { active, disabled }, className)}
        href={props.href || '#'}
        onClick={(e: MouseEvent<HTMLAnchorElement>) => {
          e.preventDefault();
          !disabled && onClick && onClick(e);
        }}
        ref={ref}>
        {children}
      </a>
    );
  }
);

export default function MenuItem(props: UseItemProps<HTMLAnchorElement>) {
  return <BaseMenuItem {...useItem(props)} />;
}
