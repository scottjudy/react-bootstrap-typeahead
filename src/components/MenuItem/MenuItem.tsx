import cx from 'classnames';
import React, { forwardRef, HTMLAttributes, MouseEvent } from 'react';

import { useItem, UseItemProps } from '../../behaviors/item';

export interface BaseMenuItemProps extends HTMLAttributes<HTMLAnchorElement> {
  active?: boolean;
  disabled?: boolean;
  href?: string;
}

export const BaseMenuItem = forwardRef<HTMLAnchorElement, BaseMenuItemProps>(
  ({ active, children, className, disabled, onClick, ...props }, ref) => {
    return (
      <a
        {...props}
        className={cx('dropdown-item', { active, disabled }, className)}
        href={props.href || '#'}
        onClick={(e: MouseEvent<HTMLAnchorElement>) => {
          e.preventDefault();
          !disabled && onClick && onClick(e);
        }}
        ref={ref}>
        {children}
      </a>
    );
  }
);

export type MenuItemProps = UseItemProps<HTMLAnchorElement>;

export default function MenuItem(props: MenuItemProps) {
  return <BaseMenuItem {...useItem(props)} />;
}
