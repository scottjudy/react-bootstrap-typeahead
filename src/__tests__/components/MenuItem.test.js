import { mount, shallow } from 'enzyme';
import { noop } from 'lodash';
import React from 'react';

import MenuItem, { BaseMenuItem } from '../../components/MenuItem.react';
import { TestProvider } from '../helpers';

const event = {
  preventDefault: noop,
};

describe('<BaseMenuItem>', () => {
  let baseMenuItem, onClick;

  beforeEach(() => {
    onClick = jest.fn();
    baseMenuItem = shallow(
      <BaseMenuItem onClick={onClick}>
        This is a base menu item.
      </BaseMenuItem>
    );
  });

  test('renders a base menu item', () => {
    expect(baseMenuItem).toBeDefined();
    expect(baseMenuItem.type()).toBe('li');
  });

  test('renders an active base menu item', () => {
    baseMenuItem.setProps({ active: true });
    expect(baseMenuItem.hasClass('active')).toBe(true);
  });

  test('triggers an event when clicked', () => {
    baseMenuItem.find('a').simulate('click', event);
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  test('renders a disabled base menu item', () => {
    baseMenuItem.setProps({ disabled: true });
    baseMenuItem.find('a').simulate('click', event);

    expect(baseMenuItem.hasClass('disabled')).toBe(true);
    expect(onClick).toHaveBeenCalledTimes(0);
  });
});

describe('<MenuItem>', () => {
  let menuItem, onClick;

  beforeEach(() => {
    onClick = jest.fn();
    menuItem = mount(
      <TestProvider selected={[]}>
        {() => (
          <MenuItem
            onClick={onClick}
            option={{ label: 'test' }}
            position={0}>
            This is a menu item.
          </MenuItem>
        )}
      </TestProvider>
    );
  });

  test('renders a menu item', () => {
    expect(menuItem).toBeDefined();
    expect(menuItem.find('a')).toHaveLength(1);
  });

  test('changes the active state of the menu item', () => {
    expect(menuItem.hasClass('active')).toBe(false);

    menuItem.setProps({ activeIndex: 0 });
    expect(menuItem.find('a').hasClass('active')).toBe(true);
  });

  test('sets the active state if it is the only result', () => {
    expect(menuItem.hasClass('active')).toBe(false);

    menuItem.setProps({
      highlightOnlyResult: true,
      results: ['test'],
    });
    expect(menuItem.find('a').hasClass('active')).toBe(true);
  });

  test('triggers an event when clicked', () => {
    menuItem.find('a').simulate('click', event);
    expect(onClick).toHaveBeenCalledTimes(1);
  });
});
