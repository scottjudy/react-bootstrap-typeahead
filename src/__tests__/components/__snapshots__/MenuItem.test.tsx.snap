// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BaseMenuItem> renders a snapshot 1`] = `
<a
  className="dropdown-item"
  href="#"
  onClick={[Function]}
>
  This is a base menu item.
</a>
`;

exports[`<MenuItem> renders a snapshot 1`] = `
<a
  aria-selected={false}
  className="dropdown-item"
  href="#"
  id="-item-0"
  onClick={[Function]}
  onMouseDown={[Function]}
  role="option"
>
  This is a menu item.
</a>
`;
